ykrt/src/mt.rs

```diff
@@ -219,215 +217,206 @@ impl MTThread {
+                // the hot count from going backwards) which can happen if an "older" thread has
+                // been paused for a long time. Even in that case, though, we do not try endlessly.
+                let mut spinwait = SpinWait::new();
                 loop {
-                    let count = lp.number_data();
-                    if count >= self.inner.hot_threshold {
-                        if self.inner.tracer.is_some() {
-                            // This thread is already tracing another Location, so either another
-                            // thread needs to trace this Location or this thread needs to wait
-                            // until the current round of tracing has completed. Either way,
-                            // there's no point incrementing the hot count.
-                            return None;
-                        }
-                        let new_state = State::phase_tracing(Arc::clone(&self.inner.tid));
-                        if loc.compare_and_swap(lp, new_state, Ordering::Relaxed) == lp {
-                            Rc::get_mut(&mut self.inner).unwrap().tracer = Some((
-                                start_tracing(self.inner.tracing_kind),
-                                Arc::clone(&self.inner.tid),
-                            ));
-                            return None;
-                        }
-                        // We raced with another thread, which probably moved the Location from
-                        // PHASE_COUNTING to PHASE_TRACING.
-                        unsafe {
-                            Arc::from_raw(new_state.pointer_data::<u8>() as *mut u8);
-                        }
-                        return None;
-                    } else {
-                        let new_state = State::phase_counting(count + 1);
-                        let old_lp = lp;
-                        lp = loc.compare_and_swap(lp, new_state, Ordering::Relaxed);
-                        if lp == old_lp {
-                            return None;
-                        }
-                        // We raced with another thread.
-                        if lp.phase() != PHASE_COUNTING {
-                            // The other thread probably moved the Location from PHASE_COUNTING to
-                            // PHASE_TRACING.
-                            return None;
+                    match loc.compare_exchange_weak(
+                        ls,
+                        ls.with_count(count + 1),
+                        Ordering::Relaxed,
+                        Ordering::Relaxed,
+                    ) {
+                        Ok(_) => break,
+                        Err(new_ls) => {
+                            if !new_ls.is_counting() {
+                                // Another thread has probably started tracing this Location.
+                                break;
+                            }
```

Comment:
I wonder if it's worth making this `else if` to make it clear that the two cases are exclusive? Tiny nit, almost not worth thinking about.